### ISSUE TITLE
Re-consider how calling already invoked tasks should behave

### ISSUE THREAD

**Issue by kenn** [capistrano/capistrano#1686]
May be a proposal for v4.

Right now, the same task can get called only once, unless `Rake::Task[].reenable` is called explicitly.

``` ruby
namespace :deploy do
  task :restart_unicorn_twice? do
    invoke 'deploy:restart_unicorn'
    invoke 'deploy:restart_unicorn'
  end

  task :restart_unicorn_twice! do
    invoke 'deploy:restart_unicorn'
    Rake::Task['deploy:restart_unicorn'].reenable
    invoke 'deploy:restart_unicorn'
  end
end
```

``` ruby
$ cap staging deploy:restart_unicorn_twice?
00:00 deploy:restart_unicorn
      01 /etc/init.d/unicorn restart
      01 Restarting
    ✔ 01 root@12.34.56.78 6.070s

$ cap staging deploy:restart_unicorn_twice!
00:00 deploy:restart_unicorn
      01 /etc/init.d/unicorn restart
      01 Restarting
    ✔ 01 root@12.34.56.78 6.074s
      01 Restarting
    ✔ 01 root@12.34.56.78 5.936s
```

However, I believe `reenable` should be set for each task by default, because:
- I know that's how rake is designed to work, but it's an implementation detail that users of capistrano shouldn't care. Let's keep in mind that even non-ruby devs are using capistrano for deployment.
- "Run only once" semantics is useful for build tasks with dependencies, which is what rake is originally designed for, but it's almost always a surprising, unwanted behavior for deployment tools. We don't think in terms of dependency. We think in terms of execution order.
- If we DO believe that "run only once" is the right way, we should at least offer a syntax sugar for `Rake::Task[].reenable` - that's very ugly and distracting.

**Comment by leehambley** [comment]
> May be a proposal for v4.

I think it would be safe enough to put that into a "feature" release, we adhere to semver, so bumping from 3-4 will mean we've done a "rewrite".

All valid points, thanks for the contribution, if you wanted to take a shot at a patch we'd be glad to consider it. 

For whatever it's worth we could try an intermediate release that hooks the `Rake::Task[].........` family of functions and directs people here to join the discussion, and we can get a few more people's input?

**Comment by kenn** [comment]
I'm wondering if there's any people who actually wrote a task that depends on the current behavior. For them, it's going to be a breaking change - but it's also believable that there are no one.

Simplest solution would be:

```ruby
module Capistrano::DSL
  def invoke(task, *args)
    Rake::Task[task].invoke(*args)
    Rake::Task[task].reenable
  end
end
```

If we want to give an option to disable it:

```ruby
module Capistrano::DSL
  def invoke(task, reenable: true, **args)
    Rake::Task[task].invoke(args)
    Rake::Task[task].reenable if reenable
  end
end
```

But we can only support `Hash` args. I vote for always reenable, even though it's going to be a breaking change. Or we can offer a global variable to configure the behavior.

**Comment by leehambley** [comment]
That's a monkeypatch I could live with, perhaps since this has the *potential* to be a breaking change, we could rather do something to issue deprecation warnings to people who're calling `reenable` already, that it'll be unnecessary in the future as we're going to make it the default?

**Comment by kenn** [comment]
I think it's going to be a breaking change for those who depend on the current (de-dup'd) behavior, by not calling `reenable`, so deprecation warning wouldn't help them.

I can't think of any solution without overly complicate the system (track all nested task calls and detect duplicate, etc.) unfortunately. I still think there's a good chance that no once is using that "feature" though.

I think the following pseudo-code offers backward compatibility by just setting a variable. That'd be an acceptable middle ground?

```ruby
set :reenable, false # default: true

module Capistrano::DSL
  def invoke(task, *args)
    Rake::Task[task].invoke(*args)
    Rake::Task[task].reenable if fetch(:reenable) || true
  end
end
```

**Comment by mattbrictson** [comment]
I am conflicted on this one.

Over the years I've found most of my frustrations with Capistrano can be traced back to its roots in Rake. For whatever reason, certain aspects of Rake are very counterintuitive to me. Things like Capistrano's `invoke`, `before`, and `after` all occupy this weird middle ground that mixes Rake and procedural behavior.

We could, piece by piece, modify Capistrano to work more procedurally, like you suggest in this PR. But ultimately Capistrano is still very much built on Rake as a foundation. Long term, won't that make the source code harder to grok if it has conflicting modes of operation? At what point does the Rake foundation become a liability?

If we were to do a v4 rewrite, then yes, perhaps all options could be on the table, including removing Rake. But I don't think a rewrite is feasible. The community is already fragmented between v2 and v3 as it is.

I guess if we think the percentage of Capistrano users impacted by this is small, then let's go forward with this PR. We could implement the feature flag as you suggest, just to ease the transition. But eventually I would suggest we remove the flag entirely, because testing and supporting two modes of operation indefinitely is too much work.

**Comment by leehambley** [comment]
It's interesting that your opinion runs nearly 100% opposite to mine, I tend to think of problems in terms of "build systems" I don't think I've touched a project since I started my career that didn't have a `Makefile` or `Rakefile` that drove the whole thing, tests, builds, data collection, deployment, provisioning, etc, etc, etc - I seem to be wired to think of things in terms of dependency resolution systems.

I'd cherry-pick some other things of what you said, but agreed, with the fragmentation as it is, I don't think a rewrite is feasible.

I also don't think *too* many people run into this, at least we could solve that with an FAQ entry, or with a simple message fired by a monkey patch to `Rake::Task` wherever the invocation is dropped incase it's already been re-run (assuming there is such a place) that warns that this is a forthcoming change.

So far as I know we're "clean" in terms of not calling already invoked tasks more than once, I think it would tell in the verbose/trace log though, so there is scope to issue a deprecation warning at least.

**Comment by mattbrictson** [comment]
> a simple message fired by a monkey patch to Rake::Task wherever the invocation is dropped incase it's already been re-run

Good idea. That will alert people that the change is coming in a future release. If it turns out a lot of people are being annoyed by it and the change will cause widespread problems, there will still be time to back out.

**Comment by leehambley** [comment]
Quite a large method to patch, it seems - https://github.com/ruby/rake/blob/master/lib/rake/task.rb#L184

**Comment by mattbrictson** [comment]
Hmm. So I guess I'm not clear how far we're taking this.

* Are we re-enabling tasks only if they are invoked via Capistrano's `invoke` DSL method? Or are we changing Rake's execution model everywhere?
* What about prerequisites? If you invoke a task a second time, do the prerequisites of that task get re-invoked as well?

This feels like it could get messy.

**Comment by leehambley** [comment]
> Hmm. So I guess I'm not clear how far we're taking this.

Suggest we shelve this, and make sure that there's a prominent warning directing people to this issue per my last suggestion, it's a bit of a shame that Rake makes that quite an awkward monkey patch. 

We could also open a new stub issue "Decide how tasks should work..." but there's some valuable notes here now.

**Comment by mattbrictson** [comment]
How about this:

1. Re-title this issue "Re-consider how tasks should work" and add a **discuss** label.
2. Create a new issue: "Feature: Emit warning when invoke is used on a task that has already been invoked"

**Comment by leehambley** [comment]
Good shout, this is updated then, and I'll create the other issue now and link them.

**Comment by mattbrictson** [comment]
My current feeling on this is that we should limit our changes to Capistrano's DSLs method, and not alter Rake itself.

Perhaps we leave `Capistrano::DSL#invoke` unchanged (to preserve backwards compatibility), and introduce a new method:

```ruby
# Force invocation of the given task, even if it has already been run
def invoke!(task, *args)
  Rake::Task[task].reenable
  invoke(task, *args)
end
```

**Comment by sreuterle** [comment]
I think multiple calls of the same task should be possible. 
My scenario: 
I have two tasks for backsyncing - one for files, one for database. Both call the same cleanup task at the end. Furthermore I have a task which invoke both of the backsync-tasks. So I have the choice to sync files and database at once or database only. So now capistrano skips the last cleanup task which is not nice. I think if I call a task twice it should be called twice - maybe output a warning. `.reenable` seems to be a dirty hack from my point of view.

**Comment by davidalger** [comment]
@sreuterle Something I implemented in a gem I did for deploying Magento 2 was to add a reenable call to the end of the task declaration for a task which is expected to be called twice during a deployment. Perhaps doing this could be useful for you as well :) 

```ruby
    desc 'Sets proper permissions on application'
    task :permissions do
      on release_roles :all do
        within release_path do
            # ...
        end
      end
      Rake::Task['magento:setup:permissions'].reenable  ## make task perpetually callable
    end
```
source: https://github.com/davidalger/capistrano-magento2/blob/master/lib/capistrano/tasks/magento.rake#L166

**Comment by sreuterle** [comment]
Thank you @davidalger , that's quite the most proper way to integrate this. Never the less: I think `.reenable` is a dirty hack at all :-/

**Comment by mattbrictson** [comment]
@sreuterle What if Capistrano offered two different methods, `invoke` and `invoke!`, where the latter allows re-invocations? That way we can preserve backwards compatibility for the current `invoke` behavior.

**Comment by sreuterle** [comment]
@mattbrictson I didn't get why it is a matter to invoke a task twice or more often. That's what I expect when I write the invoke statement twice in my task ;-) As mentioned already: a warning should be enough.
Two different types of `invoke`/`invoke!` could be a compromise.

**Comment by mattbrictson** [comment]
>  I didn't get why it is a matter to invoke a task twice or more often.

I agree, it is counterintuitive. But Capistrano is built on Rake, and that's how Rake works. 😁 

http://stackoverflow.com/questions/577944/how-to-run-rake-tasks-from-within-rake-tasks/1290119#1290119

**Comment by davidalger** [comment]
@mattbrictson I'm personally on board with the introduction of the `invoke!` method (as outlined [here](https://github.com/capistrano/capistrano/issues/1686#issuecomment-233399891)). It is very likely IMO that there are lots of configurations out there which incorrectly have `invoke` calls in loops (I did myself by accident, and only realized it when cap 3.6 started warning me of the double invocations that did nothing). If people have invokes in loops by accident which suddenly start running on every loop iteration, very bad things could happen. That's just one example which comes to my mind since I uncovered it in a gem I published after 3.6 came out!

It's an unfortunate side-effect of relying on rake tasks for those who want to use invoke calls in a procedural fashion, but there could be plenty of work out there which relies on them being declarative as well.

Throwing my 2-cents in… :)

**Comment by koenpunt** [comment]
My use case is that I'm using [capistrano-foreman](https://github.com/koenpunt/capistrano-foreman), with an additional task for restarting single processes:

```rb
namespace :foreman do
  %w(start stop restart).each do |action|
    desc "#{action.capitalize} specific service"
    task :"#{action}-process", [:process, :number] do |t, args|
      on release_roles(fetch(:foreman_roles)) do
        sudo :"#{action}", "#{fetch(:foreman_app)}-#{args[:process]}-#{args[:number]}"
      end
    end
  end
end
```

This so I can do some sort of a rolling restart from Capistrano:

```rake
namespace :deploy do
  task :restart do
    on roles(:app) do
      within release_path do
        invoke :'foreman:restart-process', 'web', 1
        invoke :'foreman:restart-process', 'web', 2
      end
    end
  end
end
```

**Comment by word** [comment]
As long as you remember that rake is declarative, the behavior of 'invoke' is perfectly fine.  Changing it  because some folk incorrectly expect it to behave procedural doesn't sound great.   Doing so would be very confusing to those who are familiar with rake and even more confusing to those who use both!  For consistencies sake, please add another method instead and keep the existing behavior as it is.

**Comment by leehambley** [comment]
> For consistencies sake, please add another method instead and keep the existing behavior as it is.

I concur.

**Comment by michaelsauter** [comment]
I just came here because I have a Capistrano task that is structured in the same way as capistrano-rails' deploy:migrate/deploy:migrating and I accidentally invoked a task multiple times due to what is described in https://github.com/capistrano/rails/issues/189#issuecomment-264716988.

I now applied the fix that capistrano-rails applied, so the task is not invoked multiple times which stops the warning messages. So - thanks for putting those warning messages in Capistrano, that lead me to discover that something was actually wrong with my code!

As a result of this, I would also vote for not changing the behaviour of `invoke`. At least in my case it protected me from doing something stupid :)

**Comment by mattbrictson** [comment]
Thanks for the feedback, everyone. 👍 

In terms of next steps, it sounds like the prevailing opinion is to leave the behavior of `invoke` as-is. We could also add a new `invoke!` method that reenables the task, for users that need that functionality. Finally, we should update the wording of the warning.

Perhaps the new warning would be:

```ruby
$stderr.puts colors.colorize("Skipping task `#{task_name}'.", :yellow)
$stderr.puts "Capistrano tasks may only be invoked once. Task `#{task}' was previously invoked, so invoke(\"#{task_name}\") at #{file}:#{line} will be skipped."
$stderr.puts "If you really meant to run this task again, use invoke!(\"#{task_name}\")"
```

Would anyone like to contribute a PR for these changes?

**Comment by l8nite** [comment]
Just came here to see where the discussion was going. We have a bunch of custom gems that wrap up capistrano tasks for internal deployment, and I was getting the warning (probably due to some bad inheritance or dependency chains).

I'm relieved to see the default behavior won't change, and I really appreciate the warning as a means to finding these unintentional duplicate calls.

**Comment by stepheneb** [comment]
Appreciate the link to this issue when I tried to invoke a task twice.

Here's an example where I found it useful.

I'm running passenger:restart a second time because I'm persisting the database in the shared directory and I'm extending the rollback task to restore the production database that was saved in the previous releases  db/ directory.

file: lib/capistrano/tasks/deploy.rake
```
namespace :deploy do
  # save a backup copy of the production database to db/production.sqlite3.bak
  before :deploy, "db:remote_db_dump_to_backup"

  # restore the production database from db/production.sqlite3.bak
  after :rollback, :restore_db do
    invoke "db:remote_db_load_database_from_backup"
    # Would prefer to invoke this in the middle of the rollback task just
    # after deploy:symlink:release.
    # Instead I invove passenger:restart a second time after the
    # production database is restored.
    Rake::Task["passenger:restart"].reenable
    invoke "passenger:restart"
  end
end
```

**Comment by edap** [comment]
I'm keep getting this alert although enabling ` Rake::Task["puma:restart"].reenable`. This is the content of my deploy script:

```ruby
namespace :puma do
  desc 'Create Directories for Puma Pids and Socket'
  task :make_dirs do
    on roles(:app) do
      execute "mkdir #{shared_path}/tmp/sockets -p"
      execute "mkdir #{shared_path}/tmp/pids -p"
    end
  end

  before :start, :make_dirs
end

namespace :deploy do
  after :finishing, 'deploy:cleanup'
  after 'deploy:publishing', 'deploy:restart'
  after :restart, :clear_cache do
    on roles(:web), in: :groups, limit: 3, wait: 10 do
      # Here we can do anything such as:
      within release_path do
        execute :rake, 'tmp:clear'
      end
    end
  end

  desc "Make sure local git is in sync with remote."
  task :check_revision do
    on roles(:app) do
      unless `git rev-parse HEAD` == `git rev-parse origin/master`
        puts "WARNING: HEAD is not the same as origin/master"
        puts "Run `git push` to sync changes."
        exit
      end
    end
  end

  desc 'Initial Deploy'
  task :initial do
    on roles(:app) do
      before 'deploy:restart', 'puma:start'
      invoke 'deploy'
    end
  end

  desc 'Restart application'
  task :restart do
    on roles(:app), in: :sequence, wait: 5 do
      Rake::Task["puma:restart"].reenable
      invoke 'puma:restart'
    end
  end

  before :starting,     :check_revision
  after  :finishing,    :compile_assets
  after  :finishing,    :cleanup
  after  :finishing,    :restart

```

And after the deploy, puma is not started and I've to do i manually from the console.

**Comment by edap** [comment]
Sorry, my bad. This is the solution of my problem https://stackoverflow.com/questions/39810930/pumarestart-invoke-twice-but-i-only-call-it-once-on-my-deploy-app-via-capistr

**Comment by mattbrictson** [comment]
`invoke!` as described [above](https://github.com/capistrano/capistrano/issues/1686#issuecomment-276715927) was helpfully added by @Tensho in #1911 and merged as `7334409`.

**Comment by mattbrictson** [comment]
Closing since we now have an official workaround in the `invoke!` method.

**Comment by awilkening** [comment]
I'm currently using the same task to complete a series of seed tasks by iterating through an array. I don't want to have to write a new task for each of the items in the array, especially as the array grows/shrinks. I'm passing the element of the array into the task as an arg. 

I wasn't aware of this "feature" (each task may only be invoked once) until I saw the warning message. If my previous deployment hadn't failed, I would not have even checked the logs to see this message.

```
Capistrano tasks may only be invoked once. Since task `deploy:seed:seeding_roles' was previously invoked, invoke("deploy:seed:seeding_roles")...
If you really meant to run this task again, use invoke!("deploy:seed:seeding_roles")
THIS BEHAVIOR MAY CHANGE IN A FUTURE VERSION OF CAPISTRANO. Please join the conversation here if this affects you.
https://github.com/capistrano/capistrano/issues/1686
```

I hope that if the behavior of `invoke!` changes in the future, deployment will get a hard failure instead of a successful deployment with some messages that will probably go unnoticed.

**Comment by dorianmariefr** [comment]
I'm trying to set a default stage with:

```
invoke :production
```

in my Capfile, but when I do `cap -T` I get this warning

**Comment by Fjan** [comment]
@dorianmariecom Setting a default stage for a task is a bit more tricky than you'd expect, but you can do it with a small monkey patch, see https://github.com/capistrano/capistrano/issues/2072

**Comment by dorianmariecom** [comment]
@Fjan it's been a while I didn't use capistrano, I use kamal now
